### PR TITLE
WIP: New crate implementing execution and schema introspection

### PR DESCRIPTION
**Pull request by SimonSapin**
**Note:** this depends on ongoing work for https://github.com/apollographql/apollo-rs/issues/641, so for now this PR targets a temporary branch in order to only show relevant diffs.

---

This is an initial draft implementation of (trying to be) spec-compliant:

* [Execution](https://spec.graphql.org/October2021/#sec-Execution):
  - A trait abstraction for objects with asynchronously resolvable fields
  - Input coercion / type checking of variable values and argument values
  - Field collection (handling `@skip`, `@include`, and type conditions)
  - Handling of field error (collecting the relevant path, propagating `null`), raised by a resolver or otherwise
  - Result coercion / type checking of response data
  - `serde`-serializable data structures of a GraphQL response, request errors, field errors, and field error path
  - JSON data for variable values and response data is represented with `serde_json_bytes::Value` (which uses a wrapper of reference-counted [`Bytes`](https://crates.io/crates/bytes) for strings)

* [Schema introspection](https://spec.graphql.org/October2021/#sec-Schema-Introspection):
  - Removing and collecting schema introspection selections (`__schema` and `__type`) from an executable document, so the rest can be executed separately (whether with the above, or not like in Apollo Router)
  - Execution of those collected selections, returning a response that can be merged with the response for the rest of the request

See `crates/apollo-introspection/tests/doc_example.rs` for an API usage example.

Testing is minimal. Experience integrating with Apollo Router would be good to have before merging this PR. Execution is included in support of introspection and not exposed in the public API. If/when another use case comes we can expose it, with some more thought on the API design (especially for the resolver trait).

### CODE REVIEW

File: crates/apollo-introspection/tests/doc_example.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+) -> Result<Response, RequestErrorResponse> {
+    let introspection = SchemaIntrospectionQuery::split_from(&mut document, operation_name)?;
+    let operation = get_operation(&document, operation_name)?
+        .definition()
+        .clone();
+    let coerced_variable_values = VariableValues::coerce(schema, &operation, variable_values)?;
+    let response =
+        execute_non_introspection(schema, &document, operation_name, &coerced_variable_values)?;
+    let intropsection_response = introspection.execute_sync(schema, &coerced_variable_values)?;
+    Ok(response.merge(intropsection_response))
```
Comment (yanns) [inline crates/apollo-introspection/tests/doc_example.rs:25]:
```suggestion
    let introspection_response = introspection.execute_sync(schema, &coerced_variable_values)?;
    Ok(response.merge(introspection_response))
```

Comment (SimonSapin) [reply crates/apollo-introspection/tests/doc_example.rs:25]: Maybe some day we’ll name something interoperability section, but not today.

Good catch, thanks!

**Comment by SimonSapin** [comment]
> Input coercion / type checking of variable values and argument values

Apollo Router has roughly the same functionality in a function called `validate_input_value`. As part of introspection work, let’s have apollo-rs expose input coercion in a way that can be used by the Router instead of `validate_input_value`.

**Comment by SimonSapin** [comment]
Moved to https://github.com/apollographql/apollo-rs/pull/758

**Pull request closed without merging**